File: ride-sharing-app/data/ride.ts

export enum RIDESTATUS {
  SCHEDULED = 'SCHEDULED',
  COMPLETED = 'COMPLETED',
  CANCELLED = 'CANCELLED'
}
const RIDE_AMOUNT = 20;
export default class Ride {
  public id: number;
  public origin: number;
  public destination: number;
  public rideAmount: number;
  public rideStatus: RIDESTATUS;
  public noOfRiders: number;
  constructor(id: number, origin: number, destination: number, noofSeats: number){
    this.id = id;
    this.origin = origin;
    this.destination = destination;
    this.noOfRiders = noofSeats;
    this.rideStatus = RIDESTATUS.SCHEDULED;
  }

  public closeRide(isPreferredRider: boolean): number {
     this.rideStatus = RIDESTATUS.COMPLETED;
    return this.calculateRide(isPreferredRider);
  }
  private calculateRide(isPreferredRider: boolean): number {
    const multiplier = this.getMultiplier(isPreferredRider);
    return  (this.destination - this.origin) * this.noOfRiders * multiplier* RIDE_AMOUNT;
  }
  private getMultiplier(isPreferredRider: boolean) : number {
    if(this.noOfRiders >=2 && isPreferredRider){
      return 0.5;
    }
    else if(isPreferredRider){
      return 0.75;
    }
    else if(this.noOfRiders >=2){
      return 0.75;
    }
    else{
      return 1;
    }
  }
}

File: ride-sharing-app/go/ride.go
package main

import "fmt"

type Ride struct {
  Id int
  Source int
  Destination int
  RideAmount int
  RideStatus RIDESTATUS
  noOfRiders int
}

const RIDE_AMOUNT = 20
type RIDESTATUS string
const (
  SCHEDULED RIDESTATUS = "SCHEDULED"
  COMPLETED RIDESTATUS = "COMPLETED"
  WITHDRAWN RIDESTATUS = "WITHDRAWN"
)

func (ride *Ride) CloseRide(isPreferredRider bool) float32{
  ride.RideStatus = COMPLETED;
  return ride.calculateRide(isPreferredRider)
}


func getMultiplier(isPreferredRide bool, noOfRiders int) float32{
  if noOfRiders >=2 && isPreferredRide{
    return 0.5
  }else if isPreferredRide {
    return 0.75
  }else{
    return 1
  }
}

func (ride *Ride) calculateRide(isPreferredRider bool) float32{
  multiplier := getMultiplier(isPreferredRider, ride.noOfRiders)
  return (float32(ride.Destination) - float32(ride.Source) ) * float32(ride.noOfRiders) * multiplier * RIDE_AMOUNT
}
